File: backend/schemas/operaciones.py
"""Esquemas API panel operaciones (monitoreo vendedores / rutas)."""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import AliasChoices, BaseModel, ConfigDict, Field, field_validator, model_validator

EstadoConexion = Literal["activo", "atrasado", "offline"]
EstadoVisitaMapa = Literal["visitado", "pendiente", "incidencia"]


class OperacionesDashboardKpis(BaseModel):
    fecha: date
    total_clientes: int = 0
    clientes_visitados: int = 0
    clientes_pendientes: int = 0
    incidencias: int = 0
    vendedores_activos: int = 0
    vendedores_total: int = 0
    porcentaje_cumplimiento: float = 0.0
    visitas_pending_sync: int = 0
    ultima_sincronizacion: datetime | None = None
    kilometros_recorridos: float = 0.0


class OperacionesDashboardResponse(BaseModel):
    kpis: OperacionesDashboardKpis
    vendedores_resumen: list["VendedorOperacionesRow"] = Field(default_factory=list)


class GpsActual(BaseModel):
    lat: float | None = None
    lon: float | None = None
    updated_at: datetime | None = None


class VendedorOperacionesRow(BaseModel):
    codigo: str
    nombre: str
    activo: bool = True
    ruta_id: int | None = None
    estado_ruta: str | None = None
    estado_conexion: EstadoConexion = "offline"
    visitas_realizadas: int = 0
    visitas_pendientes: int = 0
    incidencias: int = 0
    porcentaje_avance: float = 0.0
    ultima_sync: datetime | None = None
    pending_sync_count: int = 0
    bateria_pct: int | None = None
    gps: GpsActual | None = None
    kilometros_recorridos: float = 0.0
    usa_heartbeat: bool = False
    conexion_red: str | None = None


class VendedoresListResponse(BaseModel):
    fecha: date
    items: list[VendedorOperacionesRow]


class VisitaTimelineItem(BaseModel):
    id: int
    cliente_id: str
    nombre_fantasia: str | None = None
    direccion: str | None = None
    comuna: str | None = None
    orden_ruta: int
    estado: str
    tipo_incidencia: str | None = None
    observacion: str | None = None
    foto_url: str | None = None
    fecha_hora_visita: datetime | None = None
    lat_visita: float | None = None
    lon_visita: float | None = None
    distancia_metros: float | None = None
    sync_status: str


class VendedorDetalleResponse(BaseModel):
    codigo: str
    nombre: str
    fecha: date
    ruta_id: int | None = None
    estado_ruta: str | None = None
    hora_inicio: datetime | None = None
    hora_fin: datetime | None = None
    porcentaje_cumplimiento: float = 0.0
    kilometros_recorridos: float = 0.0
    estado_conexion: EstadoConexion = "offline"
    ultima_sync: datetime | None = None
    timeline: list[VisitaTimelineItem] = Field(default_factory=list)
    incidencias: list[VisitaTimelineItem] = Field(default_factory=list)


class MarcadorMapa(BaseModel):
    visita_id: int
    cliente_id: str
    nombre_fantasia: str | None = None
    lat: float
    lon: float
    estado: EstadoVisitaMapa
    vendedor: str
    tipo_incidencia: str | None = None


class VendedorUbicacionMapa(BaseModel):
    codigo: str
    nombre: str
    lat: float
    lon: float
    estado_conexion: EstadoConexion
    updated_at: datetime | None = None


class RutaMapaResponse(BaseModel):
    fecha: date
    ruta_id: int
    vendedor: str
    vendedor_nombre: str | None = None
    marcadores: list[MarcadorMapa] = Field(default_factory=list)
    vendedor_ubicacion: VendedorUbicacionMapa | None = None


class IncidenciaRow(BaseModel):
    id: int
    ruta_id: int
    vendedor: str
    vendedor_nombre: str | None = None
    cliente_id: str
    nombre_fantasia: str | None = None
    comuna: str | None = None
    tipo_incidencia: str | None = None
    observacion: str | None = None
    foto_url: str | None = None
    tiene_foto: bool = False
    fecha_hora_visita: datetime | None = None
    sync_status: str


class IncidenciasListResponse(BaseModel):
    fecha: date
    total: int
    items: list[IncidenciaRow]


class OperacionesMetricasResponse(BaseModel):
    fecha: date
    dashboard: OperacionesDashboardKpis
    por_vendedor: list[VendedorOperacionesRow] = Field(default_factory=list)


class HeartbeatRequest(BaseModel):
    """Telemetría desde app móvil (POST /operaciones/heartbeat)."""

    model_config = ConfigDict(populate_by_name=True, extra="ignore")

    vendedor_id: str = Field(
        ...,
        min_length=1,
        max_length=64,
        validation_alias=AliasChoices("vendedor_id", "vendedorId"),
    )
    timestamp: datetime
    lat: float | None = None
    lng: float | None = None
    bateria: int | None = Field(None, ge=0, le=100)
    conexion: str | None = Field(None, max_length=32)
    pendientes: int | None = Field(None, ge=0)
    app_version: str | None = Field(
        None,
        max_length=64,
        validation_alias=AliasChoices("app_version", "appVersion"),
    )
    dispositivo: str | None = Field(None, max_length=128)

    @field_validator("vendedor_id", mode="before")
    @classmethod
    def _strip_vendedor(cls, v: object) -> object:
        if isinstance(v, str):
            return v.strip()
        return v


class TelemetryAckResponse(BaseModel):
    """ACK telemetría móvil (heartbeat, gps_track)."""

    model_config = ConfigDict(populate_by_name=True)

    ack: bool = True
    server_timestamp: datetime | None = None
    insertados: int | None = None


# Alias retrocompatible
HeartbeatAckResponse = TelemetryAckResponse


class GpsTrackPunto(BaseModel):
    """Un punto dentro de ``puntos[]`` (formato batch app móvil)."""

    model_config = ConfigDict(populate_by_name=True, extra="ignore")

    lat: float
    lon: float | None = Field(default=None, validation_alias=AliasChoices("lon", "lng"))
    lng: float | None = Field(default=None, validation_alias=AliasChoices("lng", "lon"))
    accuracy_m: float | None = Field(
        None,
        validation_alias=AliasChoices("accuracy_m", "accuracy"),
    )
    speed: float | None = None
    timestamp: datetime | None = None

    def lng_efectivo(self) -> float:
        if self.lon is not None:
            return float(self.lon)
        if self.lng is not None:
            return float(self.lng)
        raise ValueError("Cada punto requiere lon o lng")


class GpsTrackRequest(BaseModel):
    """
  POST /operaciones/gps_track — formato legacy (lat/lng) o batch (puntos[]).
    """

    model_config = ConfigDict(populate_by_name=True, extra="ignore")

    vendedor_id: str = Field(
        ...,
        min_length=1,
        max_length=64,
        validation_alias=AliasChoices("vendedor_id", "vendedorId"),
    )
    timestamp: datetime
    session_id: str | None = Field(
        None,
        max_length=128,
        validation_alias=AliasChoices("session_id", "sessionId"),
    )
    point_ids: list[str] = Field(default_factory=list)
    puntos: list[GpsTrackPunto] = Field(default_factory=list)
    lat: float | None = None
    lon: float | None = Field(default=None, validation_alias=AliasChoices("lon", "lng"))
    lng: float | None = Field(default=None, validation_alias=AliasChoices("lng", "lon"))
    accuracy: float | None = Field(
        None,
        validation_alias=AliasChoices("accuracy", "accuracy_m"),
    )
    speed: float | None = None
    battery: int | None = Field(
        None,
        ge=0,
        le=100,
        validation_alias=AliasChoices("battery", "bateria", "bateria_pct"),
    )
    app_version: str | None = Field(
        None,
        max_length=64,
        validation_alias=AliasChoices("app_version", "appVersion"),
    )

    @field_validator("vendedor_id", mode="before")
    @classmethod
    def _strip_vendedor_gps(cls, v: object) -> object:
        if isinstance(v, str):
            return v.strip()
        return v

    def lng_efectivo(self) -> float:
        if self.lon is not None:
            return float(self.lon)
        if self.lng is not None:
            return float(self.lng)
        raise ValueError("lat/lng requeridos en formato single")

    @model_validator(mode="after")
    def _formato_single_o_batch(self) -> "GpsTrackRequest":
        if self.puntos:
            return self
        if self.lat is not None and (self.lng is not None or self.lon is not None):
            return self
        raise ValueError("Se requiere lat/lng (formato single) o puntos[] (formato batch)")


GeorefEstado = Literal["pendiente", "capturada", "aplicada"]


class ClienteGeorefRow(BaseModel):
    """Fila de georef operacional (view o rutero)."""

    cliente_codigo: str
    cliente_nombre: str
    vendedor_codigo: str
    ruta_id: int = Field(description="PK bsale.rutero.id")
    direccion: str | None = None
    comuna: str | None = None
    lat: float | None = None
    lon: float | None = None
    georef_estado: GeorefEstado | str = "pendiente"
    georef_actualizada_at: datetime | None = None
    georef_actualizada_por: str | None = None


class GeorefResumen(BaseModel):
    total: int = 0
    pendientes: int = 0
    capturados: int = 0
    aplicados: int = 0


class GeorefPendientesDebug(BaseModel):
    """Diagnóstico temporal (?debug=true)."""

    total_sql: int = Field(
        description="Filas en bsale.v_clientes_sin_georef con vendedor_codigo (vista legacy)"
    )
    total_post_filtro: int = Field(
        description="Filas devueltas tras filtro rutero (georef + vendedor ruta + dedupe)"
    )
    duplicados: int = Field(
        description="Filas rutero extra antes de DISTINCT ON (bsale_id)"
    )


class GeorefPendientesResponse(BaseModel):
    total: int
    items: list[ClienteGeorefRow] = Field(default_factory=list)
    resumen: GeorefResumen = Field(default_factory=GeorefResumen)
    debug: GeorefPendientesDebug | None = None


class GeorefActualizarRequest(BaseModel):
    """Captura GPS desde app móvil."""

    model_config = ConfigDict(populate_by_name=True, extra="ignore")

    ruta_id: int = Field(..., ge=1, validation_alias=AliasChoices("ruta_id", "rutero_id"))
    lat: float
    lon: float = Field(validation_alias=AliasChoices("lon", "lng"))
    vendedor_id: str = Field(
        ...,
        min_length=1,
        max_length=50,
        validation_alias=AliasChoices("vendedor_id", "vendedorId", "vendedor"),
    )
    actualizada_por: str | None = Field(
        None,
        max_length=50,
        validation_alias=AliasChoices("actualizada_por", "actualizadaPor"),
    )

    @field_validator("vendedor_id", mode="before")
    @classmethod
    def _strip_v(cls, v: object) -> object:
        if isinstance(v, str):
            return v.strip()
        return v


class GeorefActualizarResponse(BaseModel):
    ack: bool = True
    ruta_id: int
    cliente_codigo: str
    lat: float
    lon: float
    georef_estado: str
    georef_actualizada_at: datetime | None = None
    georef_actualizada_por: str | None = None


class GeorefEstadoPatchRequest(BaseModel):
    """ERP: marcar aplicada o volver a pendiente."""

    ruta_id: int = Field(..., ge=1, validation_alias=AliasChoices("ruta_id", "rutero_id"))
    georef_estado: Literal["pendiente", "aplicada"]
    actualizada_por: str | None = Field(None, max_length=50)


class GeorefEstadoPatchResponse(BaseModel):
    ok: bool = True
    ruta_id: int
    cliente_codigo: str
    lat: float | None = None
    lon: float | None = None
    georef_estado: str
    georef_actualizada_at: datetime | None = None
    georef_actualizada_por: str | None = None

File: frontend/services/operaciones.ts
import { getApiBaseUrl } from "@/lib/api-base"
import { getAuthHeaders } from "@/lib/api"

const API_URL = getApiBaseUrl()

export type EstadoConexion = "activo" | "atrasado" | "offline"

export interface GpsActual {
  lat: number | null
  lon: number | null
  updated_at: string | null
}

export interface VendedorOperacionesRow {
  codigo: string
  nombre: string
  activo: boolean
  ruta_id: number | null
  estado_ruta: string | null
  estado_conexion: EstadoConexion
  visitas_realizadas: number
  visitas_pendientes: number
  incidencias: number
  porcentaje_avance: number
  ultima_sync: string | null
  pending_sync_count: number
  bateria_pct: number | null
  gps: GpsActual | null
  kilometros_recorridos: number
  usa_heartbeat?: boolean
  conexion_red?: string | null
}

export interface OperacionesDashboardKpis {
  fecha: string
  total_clientes: number
  clientes_visitados: number
  clientes_pendientes: number
  incidencias: number
  vendedores_activos: number
  vendedores_total: number
  porcentaje_cumplimiento: number
  visitas_pending_sync: number
  ultima_sincronizacion: string | null
  kilometros_recorridos: number
}

export interface OperacionesDashboardResponse {
  kpis: OperacionesDashboardKpis
  vendedores_resumen: VendedorOperacionesRow[]
}

export interface VisitaTimelineItem {
  id: number
  cliente_id: string
  nombre_fantasia: string | null
  direccion: string | null
  comuna: string | null
  orden_ruta: number
  estado: string
  tipo_incidencia: string | null
  observacion: string | null
  foto_url: string | null
  fecha_hora_visita: string | null
  lat_visita: number | null
  lon_visita: number | null
  distancia_metros: number | null
  sync_status: string
}

export interface VendedorDetalleMetricas {
  clientes_asignados: number
  visitados: number
  incidencias: number
  km_recorridos: number
  km_gps: number
  km_ruta_planificada: number
  desviacion_km: number
  primera_visita: string | null
  ultima_visita: string | null
  tiempo_activo_minutos: number | null
  gps_puntos_recibidos: number
  promedio_minutos_entre_visitas: number | null
}

export interface VendedorDetalleResponse {
  codigo: string
  nombre: string
  fecha: string
  ruta_id: number | null
  estado_ruta: string | null
  hora_inicio: string | null
  hora_fin: string | null
  porcentaje_cumplimiento: number
  kilometros_recorridos: number
  estado_conexion: EstadoConexion
  ultima_sync: string | null
  timeline: VisitaTimelineItem[]
  incidencias: VisitaTimelineItem[]
  metricas?: VendedorDetalleMetricas
}

export interface RecorridoPunto {
  orden: number
  tipo: "visita" | "incidencia" | "inicio" | "gps"
  cliente: string | null
  cliente_id: string | null
  visita_id: number | null
  lat: number
  lon: number
  timestamp: string | null
  detalle: string | null
}

export interface IntervaloEntreVisitas {
  desde_cliente: string
  hasta_cliente: string
  desde_ts: string | null
  hasta_ts: string | null
  minutos: number
}

export interface VendedorRecorridoResponse {
  vendedor_id: string
  vendedor_nombre: string | null
  fecha: string
  ruta_id: number | null
  inicio: { lat: number | null; lon: number | null; timestamp: string | null; fuente: string | null } | null
  ultima_posicion: {
    lat: number | null
    lon: number | null
    timestamp: string | null
    fuente: string | null
  } | null
  puntos: RecorridoPunto[]
  linea_gps: { lat: number; lon: number; timestamp: string | null }[]
  linea_heartbeat: { lat: number; lon: number; timestamp: string | null }[]
  km_recorridos: number
  km_gps: number
  km_ruta_planificada: number
  desviacion_km: number
  intervalos_entre_visitas: IntervaloEntreVisitas[]
  promedio_minutos_entre_visitas: number | null
  metricas: VendedorDetalleMetricas
}

export interface MapaGlobalVendedor {
  codigo: string
  nombre: string
  lat: number
  lon: number
  color: string
  estado_conexion: EstadoConexion
  ultima_sync: string | null
  bateria_pct: number | null
  visitas_realizadas: number
  incidencias: number
  km_gps: number
}

export interface MapaGlobalResponse {
  fecha: string
  vendedores: MapaGlobalVendedor[]
}

export function getOperacionesMapaGlobal(fecha?: string) {
  return operacionesFetch<MapaGlobalResponse>("/mapa-global", { fecha })
}

export interface GeorefHistorialRow {
  id: number
  ruta_id: number
  estado_anterior: string | null
  estado_nuevo: string
  lat: number | null
  lon: number | null
  usuario: string | null
  fecha: string
  motivo: string | null
}

export function getOperacionesGeorefHistorial(rutaId: number) {
  return operacionesFetch<{ ruta_id: number; items: GeorefHistorialRow[] }>(
    `/georef-historial/${rutaId}`,
  )
}

export function getOperacionesVendedorRecorrido(codigo: string, fecha?: string) {
  return operacionesFetch<VendedorRecorridoResponse>(
    `/vendedor/${encodeURIComponent(codigo)}/recorrido`,
    { fecha },
  )
}

export interface MarcadorMapa {
  visita_id: number
  cliente_id: string
  nombre_fantasia: string | null
  lat: number
  lon: number
  estado: "visitado" | "pendiente" | "incidencia"
  vendedor: string
  tipo_incidencia: string | null
}

export interface RutaMapaResponse {
  fecha: string
  ruta_id: number
  vendedor: string
  vendedor_nombre: string | null
  marcadores: MarcadorMapa[]
  vendedor_ubicacion: {
    codigo: string
    nombre: string
    lat: number
    lon: number
    estado_conexion: EstadoConexion
    updated_at: string | null
  } | null
}

export interface IncidenciaRow {
  id: number
  ruta_id: number
  vendedor: string
  vendedor_nombre: string | null
  cliente_id: string
  nombre_fantasia: string | null
  comuna: string | null
  tipo_incidencia: string | null
  observacion: string | null
  foto_url: string | null
  tiene_foto?: boolean
  fecha_hora_visita: string | null
  sync_status: string
}

async function operacionesFetch<T>(path: string, params?: Record<string, string | number | undefined>): Promise<T> {
  const q = new URLSearchParams()
  if (params) {
    for (const [k, v] of Object.entries(params)) {
      if (v !== undefined && v !== "") q.set(k, String(v))
    }
  }
  const qs = q.toString()
  const url = `${API_URL}/operaciones${path}${qs ? `?${qs}` : ""}`
  const res = await fetch(url, { headers: getAuthHeaders() })
  if (!res.ok) {
    const text = await res.text()
    throw new Error(text || `HTTP ${res.status}`)
  }
  return res.json() as Promise<T>
}

export function localIsoDate(d = new Date()): string {
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, "0")
  const day = String(d.getDate()).padStart(2, "0")
  return `${y}-${m}-${day}`
}

export function getOperacionesDashboard(fecha?: string) {
  return operacionesFetch<OperacionesDashboardResponse>("/dashboard", { fecha })
}

export function getOperacionesVendedores(fecha?: string) {
  return operacionesFetch<{ fecha: string; items: VendedorOperacionesRow[] }>("/vendedores", { fecha })
}

export function getOperacionesVendedor(codigo: string, fecha?: string) {
  return operacionesFetch<VendedorDetalleResponse>(`/vendedor/${encodeURIComponent(codigo)}`, { fecha })
}

export function getOperacionesRuta(rutaId: number) {
  return operacionesFetch<RutaMapaResponse>(`/ruta/${rutaId}`)
}

export function getOperacionesIncidencias(opts?: { fecha?: string; vendedor?: string; limit?: number }) {
  return operacionesFetch<{ fecha: string; total: number; items: IncidenciaRow[] }>("/incidencias", {
    fecha: opts?.fecha,
    vendedor: opts?.vendedor,
    limit: opts?.limit,
  })
}

export function getOperacionesMetricas(fecha?: string) {
  return operacionesFetch<{ fecha: string; dashboard: OperacionesDashboardKpis; por_vendedor: VendedorOperacionesRow[] }>(
    "/metricas",
    { fecha },
  )
}

export const OPERACIONES_POLL_MS = Number(process.env.NEXT_PUBLIC_OPERACIONES_POLL_MS || "30000")

export type GeorefEstado = "pendiente" | "capturada" | "aplicada"

export type GeorefFiltroEstado = "todas" | GeorefEstado

export interface ClienteGeorefRow {
  cliente_codigo: string
  cliente_nombre: string
  vendedor_codigo: string
  ruta_id: number
  direccion: string | null
  comuna?: string | null
  lat: number | null
  lon: number | null
  georef_estado: GeorefEstado | string
  georef_actualizada_at?: string | null
  georef_actualizada_por?: string | null
}

export interface GeorefResumen {
  total: number
  pendientes: number
  capturados: number
  aplicados: number
}

export interface GeorefPendientesResponse {
  total: number
  items: ClienteGeorefRow[]
  resumen: GeorefResumen
}

export function getOperacionesGeorefPendientes(opts?: {
  vendedor?: string
  vista?: "erp"
  estado?: GeorefFiltroEstado
}) {
  const q: Record<string, string | undefined> = {
    vendedor: opts?.vendedor,
    vista: opts?.vista,
    estado: opts?.estado ?? "todas",
  }
  return operacionesFetch<GeorefPendientesResponse>("/georef-pendientes", q)
}

export async function downloadOperacionesGeorefExport(opts?: {
  vendedor?: string
  estado?: GeorefFiltroEstado
}) {
  const q = new URLSearchParams()
  if (opts?.vendedor) q.set("vendedor", opts.vendedor)
  q.set("estado", opts?.estado ?? "todas")
  const url = `${API_URL}/operaciones/georef-export?${q}`
  const res = await fetch(url, { headers: getAuthHeaders() })
  if (!res.ok) {
    const text = await res.text()
    throw new Error(text || `HTTP ${res.status}`)
  }
  const blob = await res.blob()
  const href = URL.createObjectURL(blob)
  const a = document.createElement("a")
  a.href = href
  a.download = `georef_${opts?.estado ?? "todas"}.csv`
  a.click()
  URL.revokeObjectURL(href)
}

function parseApiErrorMessage(text: string, status: number): string {
  try {
    const j = JSON.parse(text) as { detail?: string | { msg?: string }[] }
    if (typeof j.detail === "string") return j.detail
    if (Array.isArray(j.detail) && j.detail[0]?.msg) return j.detail[0].msg
  } catch {
    /* texto plano */
  }
  return text.trim() || `HTTP ${status}`
}

export function tieneGeorefEfectiva(row: Pick<ClienteGeorefRow, "lat" | "lon">): boolean {
  if (row.lat == null || row.lon == null) return false
  if (row.lat === 0 && row.lon === 0) return false
  return Number.isFinite(row.lat) && Number.isFinite(row.lon)
}

export function coordsGeorefText(row: Pick<ClienteGeorefRow, "lat" | "lon">): string | null {
  if (!tieneGeorefEfectiva(row)) return null
  return `${row.lat},${row.lon}`
}

export function googleMapsUrl(lat: number, lon: number) {
  return `https://maps.google.com/?q=${lat},${lon}`
}

export async function patchOperacionesGeorefEstado(
  rutaId: number,
  georefEstado: "pendiente" | "aplicada",
) {
  const res = await fetch(`${API_URL}/operaciones/georef-estado`, {
    method: "PATCH",
    headers: {
      ...getAuthHeaders(),
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      ruta_id: rutaId,
      georef_estado: georefEstado,
    }),
  })
  if (!res.ok) {
    const text = await res.text()
    throw new Error(parseApiErrorMessage(text, res.status))
  }
  return res.json()
}
